include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
+                        {
+                            if(frame.isValid())
+                                frameConditionData[idx]
+                                    = m_particlesBox.getSuperCell(m_superCellIdx).getSizeLastFrame();
+                        }
+
+                        /* Same as numFramesToSkip, variable is required to workaround an nvcc compiler bug.
+                         * Without redefinition within the lambda the variable will be undefined.
+                         */
+                        constexpr uint32_t skipNFrames = (numWorkers + frameSize - 1u) / frameSize;
+                        if constexpr(skipNFrames > 1)
+                        {
+                            uint32_t const frameIdx = idx / frameSize;
+                            /* select N-th (N=frameIdx) frame from the list */
+                            for(uint32_t i = 1; i <= frameIdx && frame.isValid(); ++i)
+                            {
+                                frame = T_Order::next(m_particlesBox, frame);
+                            }
+                            if constexpr(std::is_same_v<Reverse, T_Order>)
+                                frameConditionData[idx] = frameIdx != 0 ? frameSize : frameConditionData[idx];
+                        }
+                        return frame;
+                    });
+
+                bool isOneFrameValid = true;
+                // iterate over all frames in the supercell
+                do
+                {
+                    if constexpr(std::is_same_v<T_AccessTag, detail::CallParticleFunctor>)
+                    {
+                        // loop over all particles in the frame
+                        forEachParticleInFrame(
+                            [&](lockstep::Idx const linearIdx)
+                            {
+                                auto const particleIdx = linearIdx % frameSize;
+
+                                auto particle = framePtrCtx[linearIdx][particleIdx];
+                                if constexpr(std::is_same_v<Reverse, T_Order>)
+                                {
+                                    bool const isParticle = particleIdx < frameConditionData[linearIdx];
+                                    if(!isParticle)
+                                        particle.setHandleInvalid();
+                                }
+                                else if constexpr(std::is_same_v<Forward, T_Order>)
+                                {
+                                    uint32_t const parIdxInSupercell = frameConditionData[linearIdx] + linearIdx;
+                                    bool const isParticle = parIdxInSupercell < numPartcilesInSupercell;
+                                    if(!isParticle)
+                                        particle.setHandleInvalid();
+                                }
+                                else
+                                    static_assert(!sizeof(T_Order), "Unsupported order policy");
```

Comment:
Discussed offline: could be kept as-is, but may be moved somewhere (e.g. checked even outside this function), as it is a precondition on the class template argument.